src/workerd/api/worker-loader.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <workerd/io/compatibility-date.capnp.h>
+#include <workerd/io/compatibility-date.h>
+#include <workerd/io/io-channels.h>
+#include <workerd/io/io-own.h>
+#include <workerd/io/worker.h>
+#include <workerd/jsg/setup.h>
+
+namespace workerd::api {
+
+class Fetcher;
+class DurableObjectClass;
+
+// JS stub pointing to a remote Worker loaded using WorkerLoader. This is not a stub for a specific
+// entrypoint, but instead the entire Worker, allowing the caller to call any entrypoint (and
+// specify arbitrary props).
+class WorkerStub: public jsg::Object {
+ public:
+  WorkerStub(IoOwn<WorkerStubChannel> channel): channel(kj::mv(channel)) {}
+
+  struct EntrypointOptions {
+    jsg::Optional<jsg::JsObject> props;
+
+    JSG_STRUCT(props);
+  };
+
+  jsg::Ref<Fetcher> getEntrypoint(jsg::Lock& js,
+      jsg::Optional<kj::Maybe<kj::String>> name,
+      jsg::Optional<EntrypointOptions> options);
+  jsg::Ref<DurableObjectClass> getDurableObjectClass(jsg::Lock& js,
+      jsg::Optional<kj::Maybe<kj::String>> name,
+      jsg::Optional<EntrypointOptions> options);
+
+  JSG_RESOURCE_TYPE(WorkerStub) {
+    JSG_METHOD(getEntrypoint);
+    JSG_METHOD(getDurableObjectClass);
+  }
+
+ private:
+  IoOwn<WorkerStubChannel> channel;
+};
+
+// JS interface for worker loader binding.
+class WorkerLoader: public jsg::Object {
+ public:
+  // Create a WorkerLoader backed by the given I/O channel.
+  //
+  // `compatDateValidation` will differ between workerd vs. production.
+  explicit WorkerLoader(uint channel, CompatibilityDateValidation compatDateValidation)
+      : channel(channel),
+        compatDateValidation(compatDateValidation) {}
+
+  struct Module {
+    // Exactly one must be filled in.
```

Comment:
bleh, I understand why you've structured it this way but it still feels yucky not being able to use a kj::OneOf here.